src/pyhf/paramview.py

```diff
@@ -0,0 +1,98 @@
+from . import get_backend, default_backend, events
+
+
+def index_helper(name, tensor_shape, batch_shape, par_map):
+    if isinstance(name, list):
+        return [index_helper(x, tensor_shape, batch_shape, par_map) for x in name]
+    x = list(range(int(default_backend.product(tensor_shape))))
+    indices = default_backend.reshape(x, tensor_shape)
+    parfield_slice = tuple(slice(None, None) for x in batch_shape) + (
+        par_map[name]['slice'],
+    )
+    indices = indices[parfield_slice]
+    return default_backend.tolist(indices)
+
+
+class ParamViewer(object):
+    """
+    Helper class to extract parameter data from possibly batched input
+    """
+
+    def __init__(self, tensor_shape, par_map, name):
+        self.tensor_shape = tensor_shape
+        self.batch_shape = tensor_shape[:-1]
+        # for more general batch shapes we need to revisit
+        assert len(self.batch_shape) <= 1
+        self._par_map = par_map
+        self._index_selection = index_helper(
+            name, tensor_shape, self.batch_shape, self._par_map
+        )
+
+        if self._index_selection:
+            cat = default_backend.astensor(
+                default_backend.concatenate(self._index_selection, axis=-1), dtype='int'
+            )
+            if self.batch_shape:
+                self._indices_concatenated = default_backend.einsum('ij->ji', cat)
```

Comment:
not fully clear to me why a transpose is needed if there's batching vs no batching.

Reply:
there is a switch in dimension order between `index_selection` and `indices_concatenated` (and thus the return type of `.get()` in order to remove superfluous ops in the modifier application.

Added a comment